File: src/main/java/com/github/bjoernpetersen/javafxplayback/JavaFxPlaybackFactory.java
package com.github.bjoernpetersen.javafxplayback;

import com.github.bjoernpetersen.jmusicbot.InitStateWriter;
import com.github.bjoernpetersen.jmusicbot.config.Config;
import com.github.bjoernpetersen.jmusicbot.config.Config.Entry;
import com.github.bjoernpetersen.jmusicbot.platform.Platform;
import com.github.bjoernpetersen.jmusicbot.platform.Support;
import com.github.bjoernpetersen.jmusicbot.playback.Playback;
import com.github.bjoernpetersen.jmusicbot.playback.PlaybackFactory;
import com.github.bjoernpetersen.jmusicbot.playback.included.Mp3PlaybackFactory;
import com.github.bjoernpetersen.jmusicbot.playback.included.WavePlaybackFactory;
import com.github.zafarkhaja.semver.Version;
import java.io.File;
import java.io.IOException;
import java.util.Collection;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import javafx.scene.media.MediaException;
import javax.annotation.Nonnull;
import javax.sound.sampled.UnsupportedAudioFileException;

public class JavaFxPlaybackFactory implements Mp3PlaybackFactory, WavePlaybackFactory,
    com.github.bjoernpetersen.mp3Playback.Mp3PlaybackFactory {

  public JavaFxPlaybackFactory() {
  }

  @Nonnull
  @Override
  public String getReadableName() {
    return "JavaFX PlaybackFactory";
  }

  @Nonnull
  @Override
  public Support getSupport(@Nonnull Platform platform) {
    switch (platform) {
      case LINUX:
      case WINDOWS:
        return Support.YES;
      case UNKNOWN:
        return Support.MAYBE;
      case ANDROID:
      default:
        return Support.NO;
    }
  }

  @Nonnull
  @Override
  public List<Entry> initializeConfigEntries(@Nonnull Config config) {
    return Collections.emptyList();
  }

  @Override
  public void initialize(@Nonnull InitStateWriter stateWriter) {
  }

  @Nonnull
  @Override
  public List<? extends Entry> getMissingConfigEntries() {
    return Collections.emptyList();
  }

  @Override
  public void destructConfigEntries() {
  }

  @Nonnull
  @Override
  public Version getMinSupportedVersion() {
    return Version.forIntegers(0, 14, 0);
  }

  @Nonnull
  @Override
  public Playback createPlayback(@Nonnull File inputFile) throws UnsupportedAudioFileException, IOException {
    try {
      return new JavaFxPlayback(inputFile);
    } catch (MediaException e) {
      if (e.getType() == MediaException.Type.MEDIA_UNSUPPORTED) {
        throw new UnsupportedAudioFileException(e.getMessage());
      } else {
        throw new IOException(e);
      }
    }
  }

  @Nonnull
  @Override
  public Collection<Class<? extends PlaybackFactory>> getBases() {
    Set<Class<? extends PlaybackFactory>> bases = new HashSet<>();
    bases.add(Mp3PlaybackFactory.class);
    bases.add(WavePlaybackFactory.class);
    bases.add(com.github.bjoernpetersen.mp3Playback.Mp3PlaybackFactory.class);
    // theoretically more are possible. Just need to create marker interfaces.
    return bases;
  }

  @Override
  public void close() {
  }
}

File: src/main/kotlin/net/bjoernpetersen/javafxplayback/JavaFxPlaybackFactory.kt
package net.bjoernpetersen.javafxplayback

import javafx.scene.media.MediaException
import net.bjoernpetersen.musicbot.api.config.Config
import net.bjoernpetersen.musicbot.spi.plugin.Playback
import net.bjoernpetersen.musicbot.spi.plugin.management.InitStateWriter
import net.bjoernpetersen.musicbot.spi.plugin.predefined.Mp3PlaybackFactory
import net.bjoernpetersen.musicbot.spi.plugin.predefined.UnsupportedAudioFileException
import net.bjoernpetersen.musicbot.spi.plugin.predefined.WavePlaybackFactory
import java.io.File
import java.io.IOException

class JavaFxPlaybackFactory : Mp3PlaybackFactory, WavePlaybackFactory {

    override val name: String = "JavaFX"
    override val description: String = "Plays MP3 or Wave songs using JavaFX's included feature"

    override fun createConfigEntries(config: Config): List<Config.Entry<*>> {
        return emptyList()
    }

    override fun createSecretEntries(secrets: Config): List<Config.Entry<*>> {
        return emptyList()
    }

    override fun createStateEntries(state: Config) {}

    override fun initialize(initStateWriter: InitStateWriter) {}

    @Throws(UnsupportedAudioFileException::class, IOException::class)
    override fun createPlayback(inputFile: File): Playback {
        try {
            return JavaFxPlayback(inputFile)
        } catch (e: MediaException) {
            if (e.type == MediaException.Type.MEDIA_UNSUPPORTED) {
                throw UnsupportedAudioFileException(e.message)
            } else {
                throw IOException(e)
            }
        }
    }

    override fun close() {}
}
